### PR TITLE
Polkadot v0.9.37

### PR DESCRIPTION
**Pull request by NunoAlexandre**
Update to Polkadot v0.9.37

A small but noteworthy change is that, from now on, extrinsic are explicitly indexed with `#[pallet::call_index(_)]`, something we need to start doing for new extrinsic in the future and be mindful of not changing indexes, as doing so may break some integration (mostly an XCM concern but still worth keeping it in mind)

❗ Blocked by https://github.com/centrifuge/centrifuge-chain/pull/1233#pullrequestreview-1360480189

### CODE REVIEW

**Comment by mikiquantum** [comment]
@NunoAlexandre Some clean up is needed and another pass over the command & service on monday. But it compiles now.

**Comment by NunoAlexandre** [comment]
> @NunoAlexandre Some clean up is needed and another pass over the command & service on monday. But it compiles now.

You rule buddy, thanks ✌🏻

File: pallets/nft/src/lib.rs
Comment (NunoAlexandre) [reply pallets/nft/src/lib.rs:161]: It's a new Substrate thing. It requires every extrinsic to have an explicit index. This index is used - probably amongst other things - to encode/decode calls.

Comment (NunoAlexandre) [reply pallets/nft/src/lib.rs:161]: Note that all of these things could be done without `#[pallet::call_index(_)]`, but then it would be implicit. So the idea is to make extrinsic indexing explicit and hopefully harder / more obvious to mess up.

Comment (lemunozm) [reply pallets/nft/src/lib.rs:161]: Oh, I understand now. Thanks, both for your detailed answer! 🙌🏻

File: pallets/connectors/src/lib.rs
Comment (lemunozm) [inline pallets/connectors/src/lib.rs:243]: Minor comment:
I guess this is because of `clippy` complaints about this. Could it maybe be better to add the `Box` into the `Router` enum, nearest to the big variant that causes this? This way, if tomorrow we have a simpler router, we do not need to allocate it.

Comment (NunoAlexandre) [reply pallets/connectors/src/lib.rs:243]: Good one, will do ✌🏻

**Comment by branan** [comment]
This looks good to me (will approve once we're ready to merge it).

My only concern is the changes to the XCM tests - do we understand what changed there (some difference in how XCM fees are calculated orsomething?). Seeing a bunch of balances in tests changed without comments makes me nervous.

**Comment by NunoAlexandre** [comment]
> This looks good to me (will approve once we're ready to merge it).
> 
> My only concern is the changes to the XCM tests - do we understand what changed there (some difference in how XCM fees are calculated orsomething?). Seeing a bunch of balances in tests changed without comments makes me nervous.

Thanks for raising this concern! Yeah, what often happens at every Polkadot update is that the weights change so slightly or the base xcm weight values and that impacts the hardcoded expected values in our tests. Note that most other projects just tests that the final balance after some xcm operations is `initialBalance - change - fee`, which would mean I would never have to touch the tests again. The reason why I hardcoded the balances was precisely to catch an unexpected big change to some Substrate/Polkadot internal value that would have a big impact on those values and have that go unnoticed. So if you look at the diff on those values, the change is fractional. So there's nothing to be nervous about.

Let me know if you'd prefer a different approach 👍

File: pallets/connectors/src/lib.rs
Comment (NunoAlexandre) [reply pallets/connectors/src/lib.rs]: Done, it was the `VersionedMultiLocation` that needed to be boxed 👍 thanks again

**Comment by branan** [comment]
Makes sense. I think it's fine - as long as the changes aren't too big (indicating a major change) or too small (possibly indicating a rounding issue) I'm happy to see us give them a quick manual review.

It might be good to add a comment to the tests about why we have hard-coded values (basically what you said in your response above). But otherwise I'm happy with the status quo

**Comment by branan** [comment]
From what I'm seeing in #1252, I think that the lint, doc, and integration failures here are all legit.@NunoAlexandre If you can get those cleaned up I think we can land this tomorrow? Everything we need to cut the release has landed in `main` at this point, so we can move forward.

File: .github/workflows/build-nix.yml
Comment (mustermeiszer) [inline .github/workflows/build-nix.yml:2]: Why do we depend on those new features? Substrate?

File: Cargo.toml
Comment (mustermeiszer) [inline Cargo.toml:70]: Did Substrate also upgrade?

File: pallets/bridge-mapping/Cargo.toml
Comment (mustermeiszer) [inline pallets/bridge-mapping/Cargo.toml:1]: Why is this completely new?

File: src/command.rs
Comment (mustermeiszer) [inline src/command.rs:484]: Where is this from?

**Review by mustermeiszer — COMMENTED** [review]
There are a lot of changes to the service here. I would like to understand them better before we merge that.

File: .github/workflows/build-nix.yml
Comment (NunoAlexandre) [reply .github/workflows/build-nix.yml:2]: See @wischli's comment on that thread: https://github.com/NixOS/nixpkgs/pull/217232#issuecomment-1491319021

File: Cargo.toml
Comment (NunoAlexandre) [reply Cargo.toml:70]: This is a `paritytech/` crate and it has been updated to `0.16.2`: https://crates.io/crates/jsonrpsee. Polkadot therefore also updated and a few other parachain projects I checked as well (such as Moonbeam)

File: pallets/bridge-mapping/Cargo.toml
Comment (NunoAlexandre) [reply pallets/bridge-mapping/Cargo.toml:1]: Thanks for the catch! I believe I have deleted this pallet in the past which was not used anymore and ,not sure how exactly, this file was brought back to life. Deleting it.

File: pallets/connectors/Cargo.toml
Comment (NunoAlexandre) [reply pallets/connectors/Cargo.toml:117]: I was having an issue compiling this pallet and ran the `subalfred` tool and found this feature was missing here so I added it

Comment (NunoAlexandre) [reply pallets/connectors/Cargo.toml:117]: I also believe it fixed a compilation error but it's been a few weeks, don't know for sure atm.

File: src/service.rs
Comment (NunoAlexandre) [reply src/service.rs:359]: I am not sure. Maybe I copied it over from another project while trying to follow the way they have adapted to these changes in the service or maybe Miguel wrote it when he took over the work around the service. I feel very foreign to the whole service side of the chain; is this something you feel comfortable with? Cause maybe it's a good idea if you get yours hands on it and/or educate me along

File: src/service.rs
Comment (NunoAlexandre) [reply src/service.rs:1]: I'm pretty sure that's the stuff Miguel worked on, I don't know the answer without digging further into this. If you got some time and will please do 🙏

File: src/command.rs
Comment (NunoAlexandre) [reply src/command.rs:434]: Basically the `relay_chain_rpc_urls`:

``` rust
pub fn collator_options(&self) -> CollatorOptions {
  CollatorOptions { relay_chain_rpc_urls: self.relay_chain_rpc_urls.clone() }
}
```

File: src/service.rs
Comment (thea-leake) [inline src/service.rs:64]: ❤️ much cleaner/quicker to read generics below with these!

File: src/service.rs
Comment (thea-leake) [inline src/service.rs:738]: 👍 on updating arg order grouping to match more similar arg types.

**Review by thea-leake — COMMENTED** [review]
LGTM!

File: ci/script.sh
Comment (NunoAlexandre) [reply ci/script.sh:51]: Thanks for the question! I've added this allowance rule because clippy was failing in a `paramter_type` `const` that was set to `0` or to `1` based on a boolean condition; Clippy has the opinion that cases like that should be converted from `if cond { 0 } else { 1 }` to `uintT::from(cond)` (see https://github.com/centrifuge/centrifuge-chain/actions/runs/4617361181/jobs/8163526796#step:7:989).

The issue is that we can't do `uint::from` for a `const` AND that if we change the values from 0/1 (for the true/false flows) to, say, 0/2, then this suggestion wouldn't work anymore and we would need to change the code again.

By ignoring this clippy rule we have that code looking the same across runtimes and we won't need to change it if we need to change any of the values. 

Hope that makes sense!

File: runtime/integration-tests/src/xcm/kusama/tests/transfers.rs
Comment (NunoAlexandre) [reply runtime/integration-tests/src/xcm/kusama/tests/transfers.rs:61]: Thanks to @branan for querying me on this and suggesting adding it 🙏

File: ci/script.sh
Comment (thea-leake) [reply ci/script.sh:51]: Ah ok, makes sense, thanks for the explanation!

**Review by mustermeiszer — APPROVED** [review]
@NunoAlexandre I have no time to dig into the service and am out the next 8 days, so I am approving this one to not letting it linger around for too long.

File: src/service.rs
Comment (branan) [reply src/service.rs:1]: The `ParachainBlockImport` handles some of the logic around preventing clients from considering forks that haven't been backed by the relay chain. I'm not sure when it was introduced, but cumulus was updated in paritytech/cumulus#1559 to make it mandatory through the use of a marker trait.

**Review by thea-leake — APPROVED** [review]

**Review by mustermeiszer — APPROVED** [review]
Re-approve

**Review by branan — APPROVED** [review]

**Review by offerijns — APPROVED** [review]

**Pull request merged**
